src/Bus/MongoBatchRepository.php

```diff
@@ -0,0 +1,276 @@
+use DateTimeInterface;
+use Illuminate\Bus\Batch;
+use Illuminate\Bus\BatchFactory;
+use Illuminate\Bus\DatabaseBatchRepository;
+use Illuminate\Bus\PendingBatch;
+use Illuminate\Bus\PrunableBatchRepository;
+use Illuminate\Bus\UpdatedBatchJobCounts;
+use Illuminate\Support\Carbon;
+use MongoDB\BSON\ObjectId;
+use MongoDB\BSON\UTCDateTime;
+use MongoDB\Driver\ReadPreference;
+use MongoDB\Laravel\Collection;
+use MongoDB\Laravel\Connection;
+use MongoDB\Operation\FindOneAndUpdate;
+use Override;
+
+use function is_string;
+use function serialize;
+use function unserialize;
+
+// Extending DatabaseBatchRepository is necessary so methods pruneUnfinished and pruneCancelled
+// are called by PruneBatchesCommand
+class MongoBatchRepository extends DatabaseBatchRepository implements PrunableBatchRepository
+{
+    private Collection $collection;
+
+    public function __construct(
+        BatchFactory $factory,
+        Connection $connection,
+        string $collection,
+    ) {
+        $this->collection = $connection->getCollection($collection);
+
+        parent::__construct($factory, $connection, $collection);
+    }
+
+    #[Override]
+    public function get($limit = 50, $before = null): array
+    {
+        if (is_string($before)) {
+            $before = new ObjectId($before);
+        }
+
+        return $this->collection->find(
+            $before ? ['_id' => ['$lt' => $before]] : [],
+            [
+                'limit' => $limit,
+                'sort' => ['_id' => -1],
+                'typeMap' => ['root' => 'array', 'document' => 'array', 'array' => 'array'],
+            ],
+        )->toArray();
+    }
+
+    #[Override]
+    public function find(string $batchId): ?Batch
+    {
+        $batchId = new ObjectId($batchId);
+
+        $batch = $this->collection->findOne(
+            ['_id' => $batchId],
+            [
+                // If the select query is executed faster than the database replication takes place,
+                // then no batch is found. In that case an exception is thrown because jobs are added
+                // to a null batch.
+                'readPreference' => new ReadPreference(ReadPreference::PRIMARY),
+                'typeMap' => ['root' => 'array', 'array' => 'array', 'document' => 'array'],
+            ],
+        );
+
+        return $batch ? $this->toBatch($batch) : null;
+    }
+
+    #[Override]
+    public function store(PendingBatch $batch): Batch
+    {
+        $result = $this->collection->insertOne([
+            'name' => $batch->name,
+            'total_jobs' => 0,
+            'pending_jobs' => 0,
+            'failed_jobs' => 0,
+            'failed_job_ids' => [],
+            // Serialization is required for Closures
+            'options' => serialize($batch->options),
+            'created_at' => $this->getUTCDateTime(),
+            'cancelled_at' => null,
+            'finished_at' => null,
+        ]);
+
+        return $this->find($result->getInsertedId());
```

Comment:
Is there a particular reason to use separate `insertOne` and `findOne` (via `$this->find()`) operations here? This looks like a candidate for `findOneAndUpdate()` with an upsert option and `RETURN_DOCUMENT_AFTER`.

Reply:
Since `_id` is the only unknown value, I'm updating this method to hydrate the `Batch` object directly from inserted data, instead of running a query.